lib/db_manager.dart

```diff
@@ -1,279 +1,266 @@
+    await database.rawDelete('delete from setsongs where set_id=? and song_id=?', [setId, songId]);
+    print("Done removing song.");
+  }
+
+  Future<void> deleteSong(int songId) async {
+    Database database = await instance.database;
+    print("Going to remove song ${songId} from database");
+    await database.rawDelete('delete from setsongs where song_id=?', [songId]);
+    await database.rawDelete('delete from songs where id=?', [songId]);
+  }
+
+  Future<void> deleteSet(int setId) async {
+    Database database = await instance.database;
+    print("Going to delete set ID ${setId}");
+    await database.rawDelete('delete from setsongs where set_id=?', [setId]);
+    await database.rawDelete('delete from sets where id=?', [setId]);
+    print("Done deleting set ID ${setId}");
+  }
+
+  Future<int> saveSet(int? setId, List<int> selectedSongIds, String setName) async {
+    Database database = await instance.database;
+    int updatedId;
+
+    if (setId == null) {
+      updatedId = await database.rawInsert('''
+    insert into sets (name) values(?);
+    ''', [setName]);
+    } else {
+      updatedId = setId;
+      await database.rawUpdate('''update sets set name=? where id=? ''',
+      [setName, setId]);
+    }
+//print("Added or updated a set with ID: ${updatedId}");
+    await database.rawDelete('delete from setsongs where set_id=?', [setId]);
+    int ordering = 0;
+    for (int songid in selectedSongIds) {
+      await database.rawInsert(
+          'insert into setsongs (set_id, song_id, ordering) values(?, ?, ?)',
+          [updatedId, songid, ordering]);
+      ordering += 1;
+    }
+
+    return updatedId;
+  }
+
+  Future<bool> doesSongExist(String path) async {
+    String filename = basename(path);
+    Database database = await instance.database;
+
+    int? num_files = Sqflite.firstIntValue(await database.rawQuery('SELECT COUNT(*) FROM songs where filename=?', [filename]));
+    if (num_files == null || num_files == 0) {
+      return false;
+    }
+    return true;
+  }
+
+  Future<Database> get database async =>
+      _database ??= await _initDatabase();
+
+  // Creates and opens the database.
+  Future <Database> _initDatabase() async {
+    Directory documentsDirectory = await getApplicationDocumentsDirectory();
+    String path = join(documentsDirectory.path, _dbName);
+    print("Going to load database from path: ${path}");
+
+    return await openDatabase(
+      path,
+      version: migrations.length,
+      onCreate: (db, version) async {
+        print("Creating new database... ${db}");
+        _runMigrations(db, 1, version);
+        print("Database tables (hopefully) created...");
+      },
+      onUpgrade: _runMigrations,
+    );
+  }
+
+  Future _runMigrations(Database db, int oldVersion, newVersion) async {
+        var batch = db.batch();
+        print("Migrate database $oldVersion -> $newVersion");
+        for (var i = oldVersion - 1; i < newVersion; i++) {

```

Comment:
```suggestion
        for (var i = oldVersion; i < newVersion; i++) {

```

Reply:
I don't think we want the '-1' here. For example, when going from version 1 -> 2, we want `oldVersion=1`, so it will run the `migrations[1]` (which in this case is `_upgrade2`)